### ISSUE TITLE
PKCS#11 work merge in the upstream

### ISSUE THREAD

**Issue by ruchi393** [OP-TEE/optee_os#4283]
This issue is created to track work ongoing to get PKCS#11 implementation upstreamed.

Currently we have different forked out versions of PKCS#11 implementation. 
The base of the work is from @etienne-lms branches and is being used as source to upstream things to mainline.
[1] https://github.com/etienne-lms/optee_os/tree/sks-to-pkcs11
[2] https://github.com/etienne-lms/optee_client/tree/sks-to-pkcs11

At the project [3] page you'll find links to the Jira instance where we keep track of work being planned by Linaro . For PKCS#11 you will find that at LOC-21 [4]. If you scroll down to "is implemented by" and expand that, then you'll see how the work has been divided into small work packages. There you can see which ones are done, the ones that we're currently working with and the ones that have not started. So, depending on your PKCS#11 needs it might or might not have the features that you need.

[3] https://www.linaro.org/projects/#LOC
[4] https://projects.linaro.org/browse/LOC-21

**Comment by ruchi393** [comment]
Updates
Current Active Tickets:
PKCS#11 Key management functions - https://projects.linaro.org/browse/LOC-12
PKCS#11 Object Functions - https://projects.linaro.org/browse/LOC-6

Plan of action as discussed in LOC monthly meeting - https://docs.google.com/document/d/15XsqgGktCrRRWiqyaz-erp_cZykwGjkBkhMD2XtUlUY/edit#heading=h.a8fqqq174gf

@ruchi393 to contribute on Object related API's for now.
@vesajaaskelainen to contribute on C_GenerateKeyPair and EC Mechanism

**Comment by vesajaaskelainen** [comment]
@ruchi393 in order to test the find api and ec keygen & sign & verify I had to also had to do get attribute support.

Are you already working on get attribute support or shall I do that?

I am currently able to do those operations with pkcs11-tool and openssl but some cleanups and verifications are still needed and the xtest additions.

I suppose those find & get attribute are pre-requisites in order for EC key operations to be practical.

**Comment by ruchi393** [comment]
I have the code of get attribute and get object size done. Will update the FindObject PR with these commits.

**Comment by ruchi393** [comment]
@vesajaaskelainen - I am planning to work next on C_SetAttributeValue() and C_CopyObject() next. 
Once these are done,I would add test cases which uses C_CopyObject() and then gets the object attributes using C_GetAttributeValue(). Does this sound ok to you ?

You mentioned that you have the code of C_SetAttributeValue() on your branch. Should I use that as reference ?

**Comment by vesajaaskelainen** [comment]
> @vesajaaskelainen - I am planning to work next on C_SetAttributeValue() and C_CopyObject() next.
> Once these are done,I would add test cases which uses C_CopyObject() and then gets the object attributes using C_GetAttributeValue(). Does this sound ok to you ?

Sounds OK.

In meanwhile I verify your updated PR.

> You mentioned that you have the code of C_SetAttributeValue() on your branch. Should I use that as reference ?

Feel free to use it. I believe there is one TODO comment there that needs to be addressed at the same time. Certificate support may have modified it a bit so may be a good idea to see if there was some changes for that.

I will test my changes on top of your PR and then can push it for look what there is currently.

**Comment by ruchi393** [comment]
@vesajaaskelainen , have you explored any conformance suits for the library testing. We can have some tests in xtest but it would be good to find some pre-existing test suite for checking the conformance.
I tried pkcs11test  (https://github.com/google/pkcs11test.git) but this doesn't seem to have all the tests in it.
SoftHSMv2 also has tests in its src/lib/tests folder but I haven't been successful in cross-compiling them. These tests look quiet elaborate.
Any other test suite you have found/are using at your end ?

**Comment by vesajaaskelainen** [comment]
> @vesajaaskelainen , have you explored any conformance suits for the library testing. We can have some tests in xtest but it would be good to find some pre-existing test suite for checking the conformance.
> I tried pkcs11test (https://github.com/google/pkcs11test.git) but this doesn't seem to have all the tests in it.
> SoftHSMv2 also has tests in its src/lib/tests folder but I haven't been successful in cross-compiling them. These tests look quiet elaborate.
> Any other test suite you have found/are using at your end ?

I don't know if you can buy some test suites -- some test suites that you can search from web seems to have gone bad but actually quite good even thou not sometimes the prettiest is the google's pkcs11 test suite. It seems to be evolving and getting in better shape. One "problem" is that it plays with xDES also in "normal" tests which kind shows a bit bad for OP-TEE as we don't have support for that.

We do have SoftHSMv2 also compiled for the device to have some functional comparison but didn't use time to run its test suites -- if I recall they had some SoftHSMv2ism in them -- which I don't blame.

Then for python world https://github.com/danni/python-pkcs11 has some tests.

For automated testing we have setup more upper level testing that uses tools like openssl s_client/s_server, libp11 and then NSS tools for tests to make sure that things work in our expected uses cases. We also run google's pkcs11tests and we also have some own code to verify some cases mainly for the extensions or if the full test logic would need too large test infra. I am yet to plug OP-TEE's upstream code base for those tests -- that goes to next year.

**Comment by vesajaaskelainen** [comment]
@ruchi393 I got a bit bored ;)... now I have also sliced up RSA key gen, RSA PKCS sign/verify/encrypt/decrypt, RSA PSS sign/verify and RSA OAEP encrypt/decrypt.

OAEP took a bit longer bit of time to verify -- and still only SHA-1 part is verified. I have been using openssl+libp11 for verification work but after digging a bit deeper it seems that openssl's engine interface cannot handle OAEP nicely as its RSA mechanism support is a bit awkward. So ended up with hardcoded SHA-1 for testing as that is what openssl's rsautl also uses (and does not seem support changing that either).

Need to go thru the EC and RSA codes thru once again and once we have find & get support merged I can start making PRs for those.

I also have random number support commits .

I suppose next would be to write xtest test battery for those.

**Comment by ruchi393** [comment]
@vesajaaskelainen , thanks a lot for taking this up. Looks like we would be able to get all the work of PKCS#11 on mainline soon.

**Comment by vesajaaskelainen** [comment]
@ruchi393 made PRs for serialization fixes.

**Comment by vesajaaskelainen** [comment]
@ruchi393 I pushed some session fixes but for these I do not have tests.

https://github.com/vesajaaskelainen/optee_os/tree/session-fixes

You mentioned that you had issue with session testing so could you try with those two commits if they help?

**Comment by ruchi393** [comment]
> @ruchi393 I pushed some session fixes but for these I do not have tests.
> 
> https://github.com/vesajaaskelainen/optee_os/tree/session-fixes
> 
> You mentioned that you had issue with session testing so could you try with those two commits if they help?

@vesajaaskelainen , thanks for the branch.

Out of the two fixes on this branch, I picked up this commit
 "ta: pkcs11: relocate shared session object db to client session"https://github.com/vesajaaskelainen/optee_os/commit/0a7a45027f78c6222c2eacf3e371112cef4e6172

This fixes the object handle sharing issue for the session objects.  Can you raise a PR for the same.

**Comment by vesajaaskelainen** [comment]
> > @ruchi393 I pushed some session fixes but for these I do not have tests.
> > https://github.com/vesajaaskelainen/optee_os/tree/session-fixes
> > You mentioned that you had issue with session testing so could you try with those two commits if they help?
> 
> @vesajaaskelainen , thanks for the branch.
> 
> Out of the two fixes on this branch, I picked up this commit
> "ta: pkcs11: relocate shared session object db to client session"[vesajaaskelainen@0a7a450](https://github.com/vesajaaskelainen/optee_os/commit/0a7a45027f78c6222c2eacf3e371112cef4e6172)
> 
> This fixes the object handle sharing issue for the session objects. Can you raise a PR for the same.

Can do -- goes to weekend thou.

**Comment by etienne-lms** [comment]
> Can do -- goes to weekend thou.

Enjoy, and thanks for the contributions.

**Comment by vesajaaskelainen** [comment]
> > @ruchi393 I pushed some session fixes but for these I do not have tests.
> > https://github.com/vesajaaskelainen/optee_os/tree/session-fixes
> > You mentioned that you had issue with session testing so could you try with those two commits if they help?
> 
> @vesajaaskelainen , thanks for the branch.
> 
> Out of the two fixes on this branch, I picked up this commit
> "ta: pkcs11: relocate shared session object db to client session"[vesajaaskelainen@0a7a450](https://github.com/vesajaaskelainen/optee_os/commit/0a7a45027f78c6222c2eacf3e371112cef4e6172)
> 
> This fixes the object handle sharing issue for the session objects. Can you raise a PR for the same.

Here: https://github.com/OP-TEE/optee_os/pull/4348

There is no that "extra" commit as even that was not complete as in example google's pkcs11 tests reveal that we still have issues what can be done from what login/session state.

It might be challening to see what are the real use cases as master is missing lots of functionality -- so I pushed my devel branch for you to observe:
https://github.com/vesajaaskelainen/optee_os/tree/devel
https://github.com/vesajaaskelainen/optee_client/tree/devel

If you take that devel branch and fix the session login then the RSA signing tests (with exception of MD5) pass. Eg. if you take that extra commit then that passes but I must say it is not complete fix and might not even be the right fix.

Please observe failures because of:
- CKR_USER_NOT_LOGGED_IN

**Comment by ruchi393** [comment]
Thanks @vesajaaskelainen for the sessions PR.

I will give your devel branches a try. One observation though about google pkcs11 tests. As I have been trying more of these, I sometime find that some of them may not be absolutely correct as per the specification. The specification is open to interpretation at lot of places.

**Comment by vesajaaskelainen** [comment]
Sessions PR continues in own. 

But what PR would you like to see next?

I can get random stuff -- as that is easy and can allocate command ID's right after @ruchi393 set/copy objects PR.

But after that should I just go up till EC key gen? -- I will first move down some changes from the commit pile to more logical places thou.

**Comment by ruchi393** [comment]
> Sessions PR continues in own.
> 
> But what PR would you like to see next?
> 
> I can get random stuff -- as that is easy and can allocate command ID's right after @ruchi393 set/copy objects PR.
> 
> But after that should I just go up till EC key gen? -- I will first move down some changes from the commit pile to more logical places thou.
Yes, I would say.

So IIUC the order would be
1. Current PR's in review get merged - set/copy object PR , session PR
@vesajaaskelainen works on following 
2. generate random patch next 
3. Generate key pair - EC key gen and sign/verify for EC

Next, for RSA key pair generation and mechanisms, would you like to work on that or do you want me to pick it up ? I think you mentioned you have the patches cleaned up and ready on your branch.

I was thinking of picking up C_Derive, C_Wrap and C_UnWrap for secret keys (symmetric keys)  parallely. Once this is done, we can extend these for EC mechanisms, then eventually for RSA.

**Comment by vesajaaskelainen** [comment]
> > Sessions PR continues in own.
> > But what PR would you like to see next?
> > I can get random stuff -- as that is easy and can allocate command ID's right after @ruchi393 set/copy objects PR.
> > But after that should I just go up till EC key gen? -- I will first move down some changes from the commit pile to more logical places thou.
> > Yes, I would say.
> 
> So IIUC the order would be
> 
> 1. Current PR's in review get merged - set/copy object PR , session PR
>    @vesajaaskelainen works on following
> 2. generate random patch next
> 3. Generate key pair - EC key gen and sign/verify for EC

OK. I pick the EC signing parts too for that.

> Next, for RSA key pair generation and mechanisms, would you like to work on that or do you want me to pick it up ? I think you mentioned you have the patches cleaned up and ready on your branch.

Let's see how far we get and adjust accordingly. I have tried to test that both EC and RSA works.

> I was thinking of picking up C_Derive, C_Wrap and C_UnWrap for secret keys (symmetric keys) parallely. Once this is done, we can extend these for EC mechanisms, then eventually for RSA.

Starting with wrapping I believe is a good thing.

Digest stuff is also something to do which is missing.

Then somewhere in between I can do the certificates PR.

**Comment by ruchi393** [comment]
> > > Sessions PR continues in own.
> > > But what PR would you like to see next?
> > > I can get random stuff -- as that is easy and can allocate command ID's right after @ruchi393 set/copy objects PR.
> > > But after that should I just go up till EC key gen? -- I will first move down some changes from the commit pile to more logical places thou.
> > > Yes, I would say.
> > 
> > 
> > So IIUC the order would be
> > 
> > 1. Current PR's in review get merged - set/copy object PR , session PR
> >    @vesajaaskelainen works on following
> > 2. generate random patch next
> > 3. Generate key pair - EC key gen and sign/verify for EC
> 
> OK. I pick the EC signing parts too for that.
> 
> > Next, for RSA key pair generation and mechanisms, would you like to work on that or do you want me to pick it up ? I think you mentioned you have the patches cleaned up and ready on your branch.
> 
> Let's see how far we get and adjust accordingly. I have tried to test that both EC and RSA works.
> 
> > I was thinking of picking up C_Derive, C_Wrap and C_UnWrap for secret keys (symmetric keys) parallely. Once this is done, we can extend these for EC mechanisms, then eventually for RSA.
> 
> Starting with wrapping I believe is a good thing.
> 
> Digest stuff is also something to do which is missing.
Yes I have Digest and remaining AES mechanisms planned for after completion of C_Derive, C_Wrap and C_UnWrap

> 
> Then somewhere in between I can do the certificates PR.
That sounds good.

Thanks for all the effort @vesajaaskelainen . Seems like we have a good plan now.

**Comment by loblik** [comment]
As I understand this most of stuff is already implemented in @etienne-lms branch. Is this upstreaming effort some major redesign? Or is it more just about reviewing everything and putting more testing into it?

Are these two storage compatible? I mean if it's possible to switch to upstream version later once it's more complete (without loosing keys).

**Comment by Emantor** [comment]
> As I understand this most of stuff is already implemented in @etienne-lms branch. Is this upstreaming effort some major redesign? Or is it more just about reviewing everything and putting more testing into it?

Both. While reviewing, some things have been refactored but the overall architecture has remained the same.

> Are these two storage compatible? I mean if it's possible to switch to upstream version later once it's more complete (without loosing keys).

This depends on the version from etiennes branch you are using. The older 3.8.0 sks branch saved the PIN as plain data in the persistent storage, the upstream TA saves a hash over the pin. Additionally the older 3.8.0 branch used different values for `CKA_CLASS` and the other attributes, so a migration from the 3.8.0 version to upstream will definitely require a migration function. The newer version on etiennes branch should have all of these fixed already, but we won't know until it is tested.

**Comment by etienne-lms** [comment]
> Are these two storage compatible

I apology @loblik, there is no guaranty that a storage used upon my branch, at any stage, is compatible with the implementation found in this OP-TEE repository.

**Comment by vesajaaskelainen** [comment]
> Are these two storage compatible? I mean if it's possible to switch to upstream version later once it's more complete (without loosing keys).

At least there are changes on how EC keys are stored (forthcoming PR has them in standard way).

**Comment by vesajaaskelainen** [comment]
In order to support multi step signing there needs to be support for calculating digest. This required one to add defines and such.

While thinking on splitting the changes to correct commits I was thinking that perhaps it would be better that changes would be part of digest support commits. And digest calculation was easy to do I though why not do that at the same time. -- well that was more work than anticipated ;) -- but as a result I do have digest codes.

@ruchi393 I hope that you did not yet start with it.

**Comment by ruchi393** [comment]
> In order to support multi step signing there needs to be support for calculating digest. This required one to add defines and such.
> 
> While thinking on splitting the changes to correct commits I was thinking that perhaps it would be better that changes would be part of digest support commits. And digest calculation was easy to do I though why not do that at the same time. -- well that was more work than anticipated ;) -- but as a result I do have digest codes.
> 
> @ruchi393 I hope that you did not yet start with it.
No, not yet. I will begin next with C_Wrap/Unwrap once C_Derive gets final.

Thanks again for all the effort in review and patches that you have been sending in.

**Comment by vesajaaskelainen** [comment]
Next batch with elliptic curve support PR's sent. 

I will assume that TA PR has enough review work so sent it before tests are ready.

**Comment by etienne-lms** [comment]
please open back. it's been a bit stalled but still in-progress.

**Comment by etienne-lms** [comment]
Dear pkcs11 contributors, for info,  currently in progress:
* https://github.com/OP-TEE/optee_os/pull/4469 ta: pkcs11 Add implementation for Wrap/Unwrap
* https://github.com/OP-TEE/optee_os/pull/4453 ta: pkcs11: Add support for elliptic curve support

review comments are welcome (mine too :)

**Comment by jforissier** [comment]
> please open back. it's been a bit stalled but still in-progress.

Re-opened and stale label removed.

**Comment by vesajaaskelainen** [comment]
keep-alive

**Comment by vesajaaskelainen** [comment]
Next PR from me would be RSA support... currently pending on fixing the key size/type pre-check before doing the PSS signature as if invalid OP-TEE panics.

**Comment by vesajaaskelainen** [comment]
I believe the RSA with PSS and OAEP would be ready for reviewing.

But as I stacked them on top of key wrapping PRs so it is probably best to wait that wrapping gets integrated first.

Pushed them to my devel branches in case someone wants to play with it.

**Comment by vesajaaskelainen** [comment]
@ruchi393 what do you plan to look at next?

My status:
- I have RSA+RSA-PSS+RSA-OAEP ready. (no wrap support)
- Certificate object (x509) support is otherwise ready but missing tests.

**Comment by ruchi393** [comment]
@vesajaaskelainen , thanks for actively contributing and reviewing on the PKCS#11 TA.

At our end, we have two things lined up :
PKCS#11 - HMAC digest family ([https://projects.linaro.org/browse/TS-16](url)
PKCS#11 - AES mechanisms [https://projects.linaro.org/browse/TS-15](url)

@vchong is currently working on these.

**Comment by vesajaaskelainen** [comment]
Pushed out PRs for RSA support.

**Comment by vesajaaskelainen** [comment]
Fyi: OpenSC project has made release of [0.22.0](https://github.com/OpenSC/OpenSC/releases/tag/0.22.0) which includes crash fixes for pkcs11-spy. So with that you should be able to run whole google's pkcs11test suite with spy enabled. You may want to upgrade your tooling.

**Comment by vesajaaskelainen** [comment]
What shall we do for DES/3DES? They are deprecated but still used by test suites mainly. Is there are need to have those for some FIPS/CC or such validation?

**Comment by vesajaaskelainen** [comment]
@jforissier could you un-pin this issue?

**Comment by vesajaaskelainen** [comment]
> > @jforissier could you un-pin this issue?
> 
> Sorry, what do you mean?

If you look at the "Issues" tab in github then you can see this issue at the top.

"ruchi393 pinned this issue on Jan 22, 2021"

So I assume there is a way to unpin it if you have enough access.

**Comment by jforissier** [comment]
> > > @jforissier could you un-pin this issue?
> > 
> > 
> > Sorry, what do you mean?
> 
> If you look at the "Issues" tab in github then you can see this issue at the top.
> 
> "ruchi393 pinned this issue on Jan 22, 2021"

Got it.
 
> So I assume there is a way to unpin it if you have enough access.

Done!
